### ISSUE TITLE
Multiple files upload with pure ruby classes 

### ISSUE THREAD

**Issue by andersonfernandes** [refile/refile#446]
Hi guys,

I need to upload multiple files with pure ruby classes, but i'm having some problems because the [solution](https://github.com/refile/refile#multiple-file-uploads) pointed on the documentation use `accepts_attachments_for`, and the object needs to be an activerecord.

Someone can help me with that?

**Comment by arogachev** [comment]
@andersonfernandes +1 for that.

In my case I have content form in modal and I want to add file attachments to it. But since form is inside of a modal, file uploading should be AJAX / jQuery based.

I don't have ActiveRecord model and don't want to save anything to database. I use https://github.com/plataformatec/mail_form and https://github.com/plataformatec/simple_form. Provided examples contains usage of attachment with ActiveRecord model. Any tips on how to setup that?

**Comment by sobrinho** [comment]
I think that won't be easy to do with the accepts_attachments_for macro right now because it relies on the AR associations for that.

But a similar macro should not be too hard to do, looking for suggestions on this.

**Comment by sobrinho** [comment]
@elmasantos will chime on this one.

**Comment by thiagoa** [comment]
Basically, we would need the client to implement an interface. For example:

```ruby
class Image
  extend Refile::Attachment

  attachment :file
end

class User
  extend Refile::Attachment

  attr_reader :images

  accepts_attachments_for :images, collection_class: Image
end
```

**Comment by thiagoa** [comment]
@sobrinho @elmasantos It would be great if we could flesh this out before implementing.

We have two dependencies on ActiveRecord:

- `reflect_on_association`
- `ActiveRecord::Relation#build`

`reflect_on_association` is used to figure out the ActiveRecord association class. With an extra `collection_class` parameter, we could easily solve the problem for the non-AR case. For example:

```ruby
accepts_attachments_for :images, collection_class: Image
```

Now onto `ActiveRecord::Relation#build`. The AR code is similar to:

```ruby
send(association_name).build(attachment => file)
```

The same functionality could be ported to a non-ActiveRecord context by doing something like:

```ruby
send(collection_name) << collection_class.new(attachment => file)
```

To make it work, we would need to trust the client to implement an interface:

- An `attr_reader :images`
- An ActiveRecord-like constructor

Take the following example:

```ruby
class Image
  extend Refile::Attachment

  attachment :file
end

class User
  extend Refile::Attachment

  attr_reader :images

  accepts_attachments_for :images, collection_class: Image

  def initialize(attributes = {})
    @images = attributes[:images] || []
  end
end
```

@sobrinho Would that be reasonable? For every other alternative I've considered, I stumbled across the need to require custom factories, a custom method to append a new image to the collection, etc. Mutating state via an attribute reader seems to pose fewer interface requirements, even though it's not 100% ideal.

We could also require a `self.build` method on `Image`, for example:

```ruby
class Image
  def self.build(attributes)
    image = new
    image.file = attributes[:file]
    image
  end

  # ...
end
```

**Comment by sobrinho** [comment]
Makes sense for me as long we keep the current behavior, i.e. make the option available but in case of AR, we set the value by default.

**Comment by thiagoa** [comment]
@sobrinho Absolutely, thank you.
